File: android/feature/admin/src/main/kotlin/com/devora/mencare/feature/admin/campaign/CampaignScreen.kt
package com.devora.mencare.feature.admin.campaign

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.Send
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.devora.mencare.core.common.formatDateShort
import com.devora.mencare.core.common.formatTime
import com.devora.mencare.core.designsystem.component.AccentButton
import com.devora.mencare.core.designsystem.component.BarAction
import com.devora.mencare.core.designsystem.component.BrandChip
import com.devora.mencare.core.designsystem.component.BrandDatePickerDialog
import com.devora.mencare.core.designsystem.component.BrandSectionLabel
import com.devora.mencare.core.designsystem.component.BrandSwitch
import com.devora.mencare.core.designsystem.component.BrandTimePickerDialog
import com.devora.mencare.core.designsystem.component.BrandTopBar
import com.devora.mencare.core.designsystem.component.DarkHeader
import com.devora.mencare.core.designsystem.component.FieldMessage
import com.devora.mencare.core.designsystem.component.FilledTextField
import com.devora.mencare.core.designsystem.component.PickerTile
import com.devora.mencare.core.designsystem.component.SegmentedTabs
import com.devora.mencare.core.designsystem.component.readableWidth
import com.devora.mencare.core.designsystem.component.validationMessageOrNull
import com.devora.mencare.core.designsystem.theme.Bone
import com.devora.mencare.core.designsystem.theme.ErrorRed
import com.devora.mencare.core.designsystem.theme.Ink
import com.devora.mencare.core.designsystem.theme.OliveLight
import com.devora.mencare.core.designsystem.theme.OliveSoft
import com.devora.mencare.core.designsystem.theme.OliveTint
import com.devora.mencare.core.designsystem.theme.OliveWood
import com.devora.mencare.core.designsystem.theme.Radii
import com.devora.mencare.core.designsystem.theme.Stone
import com.devora.mencare.core.designsystem.theme.StoneBorder
import com.devora.mencare.core.designsystem.theme.TextMuted
import com.devora.mencare.core.model.CampaignSegment
import com.devora.mencare.feature.admin.R

@Composable
fun CampaignScreen(
    onBack: () -> Unit,
    viewModel: CampaignViewModel = hiltViewModel(),
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    var picking by rememberSaveable { mutableStateOf<Picking?>(null) }

    LaunchedEffect(state.done) {
        if (state.done) onBack()
    }

    Column(modifier = Modifier.fillMaxSize().background(Bone)) {
        DarkHeader(
            contentPadding = PaddingValues(start = 0.dp, end = 0.dp, top = 0.dp, bottom = 18.dp),
        ) {
            BrandTopBar(
                title = stringResource(R.string.camp_title),
                onBack = onBack,
                trailing = { BarAction(stringResource(R.string.camp_draft), onBack, color = OliveLight) },
            )
            Row(
                modifier = Modifier.padding(horizontal = 22.dp).padding(top = 6.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    state.name.ifBlank { stringResource(R.string.camp_name) },
                    style = MaterialTheme.typography.displaySmall.copy(fontSize = 28.sp),
                    color = Bone,
                    modifier = Modifier.weight(1f),
                )
                Text(
                    stringResource(R.string.camp_reach, state.reachable, state.segmentSize),
                    style = MaterialTheme.typography.labelMedium.copy(fontSize = 11.sp),
                    color = Bone,
                    modifier = Modifier
                        .clip(RoundedCornerShape(10.dp))
                        .background(Bone.copy(alpha = 0.12f))
                        .padding(horizontal = 11.dp, vertical = 8.dp),
                )
            }
        }

        Column(
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .readableWidth()
                .imePadding()
                .padding(horizontal = 20.dp)
                .padding(top = 16.dp, bottom = 16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            FilledTextField(
                value = state.name,
                onValueChange = viewModel::setName,
                label = stringResource(R.string.camp_name),
                error = validationMessageOrNull(state.nameError),
            )

            BrandSectionLabel(stringResource(R.string.camp_segment))
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp), modifier = Modifier.fillMaxWidth()) {
                CampaignSegment.entries.forEach { segment ->
                    BrandChip(
                        text = stringResource(segmentLabel(segment)),
                        selected = state.segment == segment,
                        onClick = { viewModel.setSegment(segment) },
                        modifier = Modifier.weight(1f),
                        fill = true,
                    )
                }
            }

            FilledTextField(
                value = state.title,
                onValueChange = viewModel::setTitle,
                label = stringResource(R.string.camp_msg_title),
            )

            Column {
                Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                    BrandSectionLabel(stringResource(R.string.camp_msg_body), modifier = Modifier.weight(1f))
                    Text(
                        stringResource(R.string.camp_chars, state.body.length),
                        style = MaterialTheme.typography.bodySmall.copy(fontSize = 11.sp),
                        color = TextMuted,
                    )
                }
                Spacer(Modifier.height(8.dp))
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(16.dp))
                        .background(Bone)
                        .border(1.5.dp, OliveSoft, RoundedCornerShape(16.dp))
                        .padding(14.dp),
                ) {
                    BasicTextField(
                        value = state.body,
                        onValueChange = viewModel::setBody,
                        textStyle = MaterialTheme.typography.bodyLarge.copy(color = Ink, fontSize = 14.sp),
                        cursorBrush = SolidColor(OliveWood),
                        modifier = Modifier.fillMaxWidth().height(56.dp),
                    )
                    Spacer(Modifier.height(10.dp))
                    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        TokenChip(stringResource(R.string.camp_token_name)) { viewModel.appendToken("{{nome}}") }
                        TokenChip(stringResource(R.string.camp_token_link)) { viewModel.appendToken("{{link}}") }
                    }
                }
                FieldMessage(validationMessageOrNull(state.bodyError))
            }
            if (state.saveFailed) {
                Text(
                    stringResource(R.string.camp_error_save),
                    style = MaterialTheme.typography.bodySmall,
                    color = ErrorRed,
                )
            }

            BrandSectionLabel(stringResource(R.string.camp_preview))
            // L'anteprima è la notifica come la vede il cliente: sul telefono
            // arriva su fondo chiaro, non su una banda scura. L'icona scura è
            // il posto del marchio, come nel mockup.
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .clip(Radii.Md)
                    .background(Stone)
                    .padding(14.dp),
            ) {
                Box(
                    modifier = Modifier.size(38.dp).clip(Radii.Sm).background(Ink),
                )
                Column(Modifier.weight(1f).padding(start = 11.dp)) {
                    Text(
                        state.title.ifBlank { stringResource(R.string.camp_msg_title) },
                        style = MaterialTheme.typography.titleSmall,
                        color = Ink,
                    )
                    Text(
                        state.previewBody,
                        style = MaterialTheme.typography.bodySmall.copy(fontSize = 13.sp, lineHeight = 18.sp),
                        color = TextMuted,
                    )
                }
                Text(
                    stringResource(R.string.camp_now),
                    style = MaterialTheme.typography.labelMedium.copy(fontSize = 11.sp),
                    color = TextMuted,
                )
            }

            BrandSectionLabel(stringResource(R.string.camp_scheduling))
            SegmentedTabs(
                options = listOf(
                    stringResource(R.string.camp_send_now),
                    stringResource(R.string.camp_schedule),
                ),
                selectedIndex = if (state.scheduleLater) 1 else 0,
                onSelect = { viewModel.setScheduleLater(it == 1) },
                onDark = false,
            )
            if (state.scheduleLater) {
                Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                    PickerTile(
                        label = stringResource(R.string.camp_date),
                        value = formatDateShort(state.scheduledDate),
                        onClick = { picking = Picking.DATE },
                        modifier = Modifier.weight(1f),
                    )
                    PickerTile(
                        label = stringResource(R.string.camp_time),
                        value = formatTime(state.scheduledTime),
                        onClick = { picking = Picking.TIME },
                        modifier = Modifier.weight(1f),
                    )
                }
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clip(Radii.Md)
                        .background(Bone)
                        .border(1.5.dp, StoneBorder, Radii.Md)
                        .padding(horizontal = 16.dp, vertical = 12.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Column(Modifier.weight(1f)) {
                        Text(
                            stringResource(R.string.camp_repeat_weekly),
                            style = MaterialTheme.typography.titleMedium.copy(fontSize = 15.sp),
                            color = Ink,
                        )
                        Text(
                            stringResource(R.string.camp_stop_after, state.sendCap),
                            style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp),
                            color = TextMuted,
                        )
                    }
                    BrandSwitch(
                        checked = state.repeatWeekly,
                        onCheckedChange = viewModel::setRepeatWeekly,
                    )
                }
            }
        }

        Column(
            modifier = Modifier
                .navigationBarsPadding()
                .readableWidth()
                .padding(horizontal = 20.dp)
                .padding(bottom = 14.dp),
        ) {
            AccentButton(
                text = stringResource(
                    if (state.scheduleLater) R.string.camp_schedule_cta else R.string.camp_send_now,
                ),
                onClick = viewModel::send,
                loading = state.saving,
                height = 56.dp,
                shape = RoundedCornerShape(16.dp),
                leadingIcon = Icons.AutoMirrored.Outlined.Send,
            )
        }
    }

    when (picking) {
        Picking.DATE -> BrandDatePickerDialog(
            initial = state.scheduledDate,
            onDismiss = { picking = null },
            onConfirm = {
                viewModel.setScheduledDate(it)
                picking = null
            },
        )
        Picking.TIME -> BrandTimePickerDialog(
            initial = state.scheduledTime,
            onDismiss = { picking = null },
            onConfirm = {
                viewModel.setScheduledTime(it)
                picking = null
            },
        )
        null -> Unit
    }
}

private enum class Picking { DATE, TIME }

@Composable
private fun TokenChip(text: String, onClick: () -> Unit) {
    // Pill d'accento come le altre: fondo OliveTint pieno, testo oliva.
    Text(
        text,
        style = MaterialTheme.typography.labelMedium.copy(fontSize = 11.sp),
        color = OliveWood,
        modifier = Modifier
            .clip(Radii.Pill)
            .background(OliveTint)
            .clickable(onClick = onClick)
            .padding(horizontal = 11.dp, vertical = 8.dp),
    )
}

private fun segmentLabel(segment: CampaignSegment): Int = when (segment) {
    CampaignSegment.INATTIVI_60 -> R.string.camp_seg_inactive
    CampaignSegment.TUTTI -> R.string.camp_seg_all
    CampaignSegment.TOP_SPESA -> R.string.camp_seg_top
}

File: ios/MenCare/Features/Admin/Campaign/CampaignScreen.swift
import SwiftUI
import Observation

@MainActor
@Observable
final class CampaignViewModel {

    private let admin: AdminRepository

    var id = ""
    var name = ""
    var segment: CampaignSegment = .inattivi60
    var title = ""
    var body = ""
    var scheduleLater = false
    var scheduledDate: LocalDate = LocalDate.today().plusDays(1)
    var scheduledTime = LocalTime(10, 30)
    var repeatWeekly = false
    var sendCap = 4
    var reachable = 0
    var segmentSize = 0
    var done = false
    var sending = false
    /// Chiave del campo → messaggio: prima il tasto "Invia" con nome o
    /// messaggio vuoto usciva in silenzio e sembrava rotto.
    var fieldErrors: [String: String] = [:]
    var saveError: String?

    let state = LoadState()

    @ObservationIgnored private var submitted = false
    @ObservationIgnored private var reachTask: Task<Void, Never>?

    init(admin: AdminRepository) {
        self.admin = admin
    }

    func load() async {
        await state.run {
            // Si riprende la bozza aperta, se c'e'.
            if let draft = try await admin.campaigns().first(where: { $0.status == .draft }) {
                id = draft.id
                name = draft.name
                segment = draft.segment
                title = draft.title
                body = draft.body
                repeatWeekly = draft.repeatWeekly
                sendCap = draft.sendCap ?? 4
                scheduleLater = draft.scheduledAt != nil
                scheduledDate = draft.scheduledAt?.date ?? LocalDate.today().plusDays(1)
                scheduledTime = draft.scheduledAt?.time ?? LocalTime(10, 30)
            }
            // La copertura la conta il server: quanti riceverebbero davvero la
            // push, non un numero gonfiato per somigliare al mockup.
            let reach = try await admin.reachFor(segment)
            reachable = reach.reachable
            segmentSize = reach.size
        }
    }

    /// Push preview with merge tokens personalised for the sample client.
    var previewBody: String {
        body
            .replacingOccurrences(of: "{{nome}}", with: "Marco")
            .replacingOccurrences(of: "{{link}}", with: "mencare.app/prenota")
    }

    func refreshReach() {
        reachTask?.cancel()
        reachTask = Task {
            guard let reach = try? await admin.reachFor(segment) else { return }
            reachable = reach.reachable
            segmentSize = reach.size
        }
    }

    func setSegment(_ newSegment: CampaignSegment) {
        segment = newSegment
        refreshReach()
    }

    func fieldChanged() {
        saveError = nil
        if submitted { fieldErrors = validate() }
    }

    func appendToken(_ token: String) {
        body = String((body + token).prefix(campaignBodyLimit))
        fieldChanged()
    }

    private func validate() -> [String: String] {
        var errors: [String: String] = [:]
        errors["name"] = validateRequiredText(name).message
        errors["title"] = validateRequiredText(title).message
        let trimmedBody = body.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmedBody.isEmpty {
            errors["body"] = L("validation_required")
        } else if body.count > campaignBodyLimit {
            errors["body"] = L("validation_note_max", campaignBodyLimit)
        }
        return errors.compactMapValues { $0 }
    }

    func send() {
        guard !sending else { return }
        submitted = true
        let errors = validate()
        if !errors.isEmpty {
            fieldErrors = errors
            saveError = L("validation_form_invalid")
            return
        }
        fieldErrors = [:]
        saveError = nil
        sending = true
        Task {
            let result = await admin.saveCampaign(
                PushCampaign(
                    id: id,
                    name: name,
                    segment: segment,
                    title: title,
                    body: body,
                    scheduledAt: scheduleLater ? scheduledDate.atTime(scheduledTime) : nil,
                    repeatWeekly: repeatWeekly,
                    sendCap: repeatWeekly ? sendCap : nil,
                    reachableCount: reachable,
                    segmentSize: segmentSize,
                    status: scheduleLater ? .scheduled : .sent
                )
            )
            sending = false
            switch result {
            case .success:
                done = true
            case .failure(let error):
                // Il server valida nome, titolo e data d'invio: il suo messaggio
                // dice piu' di un generico "non riuscito".
                saveError = error.displayMessage
            }
        }
    }
}

private enum Picking: Identifiable {
    case date, time
    var id: Self { self }
}

struct CampaignScreen: View {
    @Bindable var viewModel: CampaignViewModel
    let onBack: () -> Void

    @State private var picking: Picking?

    var body: some View {
        VStack(spacing: 0) {
            DarkHeader(contentPadding: EdgeInsets(top: 0, leading: 0, bottom: 18, trailing: 0)) {
                BrandTopBar(title: L("camp_title"), onBack: onBack) {
                    BarAction(text: L("camp_draft"), action: onBack, color: .oliveLight)
                }
                HStack {
                    Text(viewModel.name.isEmpty ? L("camp_name") : viewModel.name)
                        .font(Typo.cormorant(28))
                        .foregroundStyle(Color.bone)
                    Spacer()
                    Text(L("camp_reach", viewModel.reachable, viewModel.segmentSize))
                        .font(Typo.jost(11, weight: .medium))
                        .foregroundStyle(Color.bone)
                        .padding(.horizontal, 11)
                        .padding(.vertical, 8)
                        .background(RoundedRectangle(cornerRadius: 10).fill(Color.bone.opacity(0.12)))
                }
                .padding(.horizontal, 22)
                .padding(.top, 6)
            }

            ScrollView {
                VStack(alignment: .leading, spacing: 12) {
                    FilledTextField(
                        label: L("camp_name"), text: $viewModel.name,
                        autocapitalization: .sentences,
                        error: viewModel.fieldErrors["name"], outlined: true
                    )
                    .onChange(of: viewModel.name) { viewModel.fieldChanged() }

                    BrandSectionLabel(text: L("camp_segment"))
                    HStack(spacing: 8) {
                        segmentChip(.inattivi60, L("camp_seg_inactive"))
                        segmentChip(.tutti, L("camp_seg_all"))
                        segmentChip(.topSpesa, L("camp_seg_top"))
                    }

                    FilledTextField(
                        label: L("camp_msg_title"), text: $viewModel.title,
                        autocapitalization: .sentences,
                        error: viewModel.fieldErrors["title"]
                    )
                    .onChange(of: viewModel.title) { viewModel.fieldChanged() }

                    bodyEditor

                    FormErrorBanner(message: viewModel.saveError)

                    BrandSectionLabel(text: L("camp_preview"))
                    preview

                    BrandSectionLabel(text: L("camp_scheduling"))
                    SegmentedTabs(
                        options: [L("camp_send_now"), L("camp_schedule")],
                        selectedIndex: viewModel.scheduleLater ? 1 : 0,
                        onSelect: { viewModel.scheduleLater = $0 == 1 },
                        onDark: false
                    )
                    if viewModel.scheduleLater {
                        HStack(spacing: 10) {
                            PickerTile(label: L("camp_date"), value: formatDateShort(viewModel.scheduledDate)) {
                                picking = .date
                            }
                            PickerTile(label: L("camp_time"), value: formatTime(viewModel.scheduledTime)) {
                                picking = .time
                            }
                        }
                        HStack {
                            VStack(alignment: .leading, spacing: 0) {
                                Text(L("camp_repeat_weekly"))
                                    .font(Typo.jost(15, weight: .medium))
                                    .foregroundStyle(Color.ink)
                                Text(L("camp_stop_after", viewModel.sendCap))
                                    .font(Typo.jost(12))
                                    .foregroundStyle(Color.textMuted)
                            }
                            Spacer()
                            BrandSwitch(isOn: $viewModel.repeatWeekly)
                        }
                        .padding(.horizontal, 16)
                        .padding(.vertical, 12)
                        .background(RoundedRectangle(cornerRadius: 16).fill(Color.stone))
                    }
                }
                .padding(.horizontal, 20)
                .padding(.top, 16)
                .padding(.bottom, 16)
                .readableWidth()
            }

            AccentButton(
                text: L(viewModel.scheduleLater ? "camp_schedule_cta" : "camp_send_now"),
                action: viewModel.send,
                loading: viewModel.sending,
                height: 56,
                corner: 16,
                leadingSystemImage: "paperplane"
            )
            .padding(.horizontal, 20)
            .padding(.bottom, 14)
            .readableWidth()
        }
        .background(Color.bone)
        .task { await viewModel.load() }
        .onChange(of: viewModel.done) {
            if viewModel.done { onBack() }
        }
        .sheet(item: $picking) { mode in
            switch mode {
            case .date:
                BrandDatePickerDialog(
                    initial: viewModel.scheduledDate,
                    onDismiss: { picking = nil },
                    onConfirm: { date in
                        viewModel.scheduledDate = date
                        picking = nil
                    }
                )
            case .time:
                BrandTimePickerDialog(
                    initial: viewModel.scheduledTime,
                    onDismiss: { picking = nil },
                    onConfirm: { time in
                        viewModel.scheduledTime = time
                        picking = nil
                    }
                )
            }
        }
    }

    private func segmentChip(_ segment: CampaignSegment, _ label: String) -> some View {
        BrandChip(
            text: label,
            selected: viewModel.segment == segment,
            action: { viewModel.setSegment(segment) },
            fill: true
        )
    }

    private var bodyEditor: some View {
        CounterTextField(
            label: L("camp_msg_body"),
            text: $viewModel.body,
            limit: campaignBodyLimit,
            error: viewModel.fieldErrors["body"]
        ) {
            HStack(spacing: 8) {
                tokenChip(L("camp_token_name"), token: "{{nome}}")
                tokenChip(L("camp_token_link"), token: "{{link}}")
            }
        }
        .onChange(of: viewModel.body) { viewModel.fieldChanged() }
    }

    private func tokenChip(_ label: String, token: String) -> some View {
        Button {
            viewModel.appendToken(token)
        } label: {
            Text(label)
                .font(Typo.jost(12))
                .foregroundStyle(Color.ink)
                .padding(.horizontal, 11)
                .padding(.vertical, 8)
                .background(RoundedRectangle(cornerRadius: 10).fill(Color.stone))
        }
        .buttonStyle(.plain)
    }

    private var preview: some View {
        HStack(alignment: .top, spacing: 11) {
            RoundedRectangle(cornerRadius: 10)
                .fill(Color.bone.opacity(0.14))
                .frame(width: 34, height: 34)
            VStack(alignment: .leading, spacing: 2) {
                Text(viewModel.title.isEmpty ? L("camp_msg_title") : viewModel.title)
                    .font(Typo.titleSmall)
                    .foregroundStyle(Color.bone)
                Text(viewModel.previewBody)
                    .font(Typo.jost(12))
                    .foregroundStyle(Color.bone)
            }
            Spacer()
            Text(L("camp_now"))
                .font(Typo.jost(11))
                .foregroundStyle(Color.bone.opacity(0.6))
        }
        .padding(14)
        .background(RoundedRectangle(cornerRadius: 16).fill(Color.ink))
    }
}
